gitrevise/todo.py

```diff
@@ -244,11 +244,23 @@ def edit_todos(
 
 def apply_todos(
     current: Optional[Commit],
+    original: List[Step],
     todos: List[Step],
     reauthor: bool = False,
 ) -> Commit:
-    for step in todos:
-        rebased = step.commit.rebase(current).update(message=step.message)
+    applied_old_commits = set()
+    applied_new_commits = set()
+
+    for known, step in zip(original, todos):
+        # Avoid making the user resolve the same conflict twice:
+        # When reordering commits, the final state is known.
+        applied_old_commits.add(known.commit.oid)
+        applied_new_commits.add(step.commit.oid)
+        is_known = applied_new_commits == applied_old_commits
+        known_state = known.commit if is_known else None
```

Comment:
Maybe better to compute the tree here already and call this `known_tree` (less opaque than `known_state`).
We only care about the tree, not the commit

Reply:
Makes a lot of sense.

Although I must admit, as a mere git user, I didn't know what a `tree` was until I started hacking at this.

Reply:
I think let's call it `keep_tree`. That name says not just what it is, but what it does.

Reply:
done